src/interaction/keyframerecording.cpp

```diff
@@ -0,0 +1,269 @@
+ * software and associated documentation files (the "Software"), to deal in the Software *
+ * without restriction, including without limitation the rights to use, copy, modify,    *
+ * merge, publish, distribute, sublicense, and/or sell copies of the Software, and to    *
+ * permit persons to whom the Software is furnished to do so, subject to the following   *
+ * conditions:                                                                           *
+ *                                                                                       *
+ * The above copyright notice and this permission notice shall be included in all copies *
+ * or substantial portions of the Software.                                              *
+ *                                                                                       *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,   *
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A         *
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT    *
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF  *
+ * CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE  *
+ * OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                                         *
+ ****************************************************************************************/
+
+#include <openspace/interaction/keyframerecording.h>
+
+#include <openspace/camera/camera.h>
+#include <openspace/engine/globals.h>
+#include <openspace/engine/windowdelegate.h>
+#include <openspace/navigation/navigationhandler.h>
+#include <openspace/navigation/orbitalnavigator.h>
+#include <openspace/scene/scene.h>
+#include <openspace/scene/scenegraphnode.h>
+#include <openspace/util/timemanager.h>
+#include <ghoul/filesystem/filesystem.h>
+#include <ghoul/glm.h>
+#include <algorithm>
+#include <iostream>
+
+#include "keyframerecording_lua.inl"
+
+namespace {
+    constexpr std::string_view _loggerCat = "KeyframeRecording";
+} // namespace
+
+namespace openspace::interaction {
+
+KeyframeRecording::KeyframeRecording()
+    : properties::PropertyOwner({ "KeyframeRecording", "Keyframe Recording" })
+{}
+
+void KeyframeRecording::newSequence(std::string filename) {
+    saveSequence();
+
+    _keyframes.clear();
+    _filename = std::move(filename);
+}
+
+void KeyframeRecording::addKeyframe(double sequenceTime) {
+    nlohmann::json keyframe = newKeyframe(sequenceTime);
+
+    auto it = std::find_if(
+        _keyframes.begin(),
+        _keyframes.end(),
+        [sequenceTime](const nlohmann::json& entry) {
+            return sequenceTime < static_cast<double>(entry["timestamp"]["sequence"]);
+        }
+    );
+    _keyframes.insert(it, keyframe);
+}
+
+void KeyframeRecording::updateKeyframe(int index) {
+    nlohmann::json old = _keyframes.at(index);
```

Comment:
This will crash if the user provides an `index` that is <0 or bigger than the number of keyframes